infrastructure/src/main/java/com/music/s3/S3Storage.java

```diff
@@ -0,0 +1,148 @@
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class S3Storage implements FileStorage {
+
+  private final AmazonS3 amazonS3;
+  private final TransferManager transferManager;
+  private final FileValidator fileValidator;
+
+  @Value("${aws.s3.bucket}")
+  private String bucket;
+
+  @Override
+  public FileKeys uploadTrack(TrackUpload trackUpload) {
+    validateTrackUpload(trackUpload);
+    return uploadMultipleFiles(trackUpload);
+  }
+
+  @Override
+  public InputStream getFileStream(String fileKey) {
+    try {
+      return amazonS3.getObject(bucket, fileKey).getObjectContent();
+    } catch (AmazonS3Exception e) {
+      log.error("파일 가져오기 실패. key: {}, error: {}", fileKey, e.getMessage());
+      throw new RuntimeException(); // TODO: CustomException 으로 변경
+    }
+  }
+
+  private void validateTrackUpload(TrackUpload trackUpload) {
+    MusicUpload musicUpload = trackUpload.getMusicUpload();
+    trackUpload.getLyricsUploadOptional()
+        .ifPresent(lyrics -> {
+          fileValidator.validateFileWithMetadata(
+              lyrics.getLyricsFile(),
+              lyrics.getContentType(),
+              lyrics.getOriginalFilename()
+          );
+        });
+
+    for (Map.Entry<AudioQuality, File> entry : musicUpload.getFileByAudioQuality().entrySet()) {
+      fileValidator.validateFileExists(entry.getValue());
+    }
+    fileValidator.validateContentType(musicUpload.getContentType());
+    fileValidator.validateFileName(musicUpload.getOriginalFilename());
+  }
+
+  private FileKeys uploadMultipleFiles(TrackUpload trackUpload) {
+    MusicUpload musicUpload = trackUpload.getMusicUpload();
+    LyricsUpload lyricsUpload = trackUpload.getLyricsUpload();
+
+    List<File> files = trackUpload.getFiles();
+    String directory = FileType.TRACK.getDirectory();
+
+    ObjectMetadataProvider provider = new CustomObjectMetadataProvider(musicUpload, lyricsUpload);
+    try { // AWS 권장 방식 업로드
+      MultipleFileUpload multipleFileUpload = transferManager.uploadFileList(
+          bucket,
+          directory,
+          FileUtils.createTempFile(), // 업로드 대상 파일이 공통된 기본 디렉토리 하위에 있어야 함
+          files,
+          provider
+      );
+
+      multipleFileUpload.addProgressListener(createProgressListener());
+      multipleFileUpload.waitForCompletion();
+
+      log.info("모든 파일 업로드 완료");
+      return createFileKeys(trackUpload, directory);
+    } catch (AmazonServiceException e) {
+      log.error("AWS S3 Service error : {}", e.getMessage());
+      throw new RuntimeException(); // TODO: CustomException 으로 변경
+    } catch (AmazonClientException e) {
+      log.error("AWS S3 Client error : {}", e.getMessage());
+      throw new RuntimeException(); // TODO: CustomException 으로 변경
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      log.error("File upload interrupted : {}", e.getMessage());
+      throw new RuntimeException(); // TODO: CustomException 으로 변경
```

Comment:
아래처럼 기본적으로 던져지는 e는 Exception에 같이 넣어서 넘겨주시면 좋을 것 같아요
```suggestion
      throw new RuntimeException(e); // TODO: CustomException 으로 변경
```